Switch from script elements to comments and comment elements

nateps opened the pull request:
It required a series of crazy IE hacks, but I figured out how to make comments work in all browsers.

The advantages to comments are:
- They don't affect sibling CSS selectors, while script tags do. For example, `+`, `:first-child`, and `:nth-child()` can all be messed up by using script markers.
- DOM traversal methods that only operate on elements will ignore comment markers but not script markers
- HTML comments are valid anywhere, whereas script elements are not even though they work

First, regular comments can be pulled out quickly in IE9 and all other browsers via document.createTreeWalker. This method is actually equivalent to using document.createNodeIterator on comment nodes, because they can't have children. The advantage to createTreeWalker is that it is supported in old versions of Firefox (works in 2.0; didn't bother to test in anything older), whereas createNodeIterator wasn't implemented until Firefox 3.5.

IE 6-8, of course, is what makes this exciting. First of all, legacy IE supports the `<comment>` element that like other elements can have an id and can be retrieved via document.getElementById. However, the comment element acts entirely like an HTML comment from a parsing and rendering perspective. Thus, these elements don't affect CSS selectors. (The exception is :first-child in IE7, which does not properly ignore non-element nodes of any type.) The other problems are that IE doesn't seem to allow comment nodes to be children of select elements. However, the contents of the comment become a non-rendered textNode child of the select, so it is possible to special case this particular bug and just check the childNodes of all the select elements. Finally, IE has major issues with inserting "zero-scope" nodes, including comments, via innerHTML. I figured out that somehow putting `<script></script>` keeps IE from stripping out comment elements that follow, and these empty script tags are removed even when in the middle of content inserted via innerHTML. I added a test to double check that case.

All tests passed in Mac Firefox 2.0, 3.0, 3.6, 9.0, current Chrome, Opera, Safari, and XP IE6, 7, 8, and Windows 7 IE9.

nateps committed:
range.setStart requires offset parameter in FF3.6 and Opera

nateps committed:
Switch to HTML comments for modern browsers and comment elements for IE6-8

While script tag markers work in all browsers and are easier to access
via document.getElementById, this set of optimizations and hacks makes
using comments workable. Script tag markers cause problems with sibling
CSS selectors that comments and comment elements do not affect. In
addition, comments are ignored by element based DOM traversal methods,
while script tags are not. Finally, comments are valid in all DOM
locations, so their use seems less brittle.

nateps commented:
Note, there are lots of old comments that assumed the script marker implementation, and the way that morph.checkRemoved is being called from the .html function but not by other mutators should be cleaned up. Given that the checkRemoved function has to try and get the markers, it is probably better to do this check in the mutator functions instead of getting them twice. However, these seemed out of scope of the pull request.

jamesarosen commented on lib/metamorph.js:
```diff
@@ -146,7 +180,51 @@
+
   } else {
+
+    if (document.createElement('comment').length == 0) {
```
JSHint prefers `=== 0`

ebryn commented:
@nateps I've put together a comparison of performance between your comment-based solution and the existing script tag based approach. It looks like the comment-based approach is significantly slower.

http://jsperf.com/metamorph-script-vs-comments/2

ebryn commented:
Here's another test that just tests update performance: http://jsperf.com/metamorph-script-vs-comments/3

nateps committed:
Make some performance improvements and throw errors when markers aren't found in all mutator methods

nateps committed:
Update to make IE9 work with ranges, and use same fragmentFor function for all IE versions

Unfortunately, IE9 removes comments inside of select tags, so the
select test now fails in IE9. A workaround would be to use text nodes
instead of comments inside of select elements only, but there is not a
clean way of implementing this without requiring developers to specify
whether the container is a select when calling the outerHTML function.

nateps commented:
Made some performance fixes. For some reason, jsperf returns very different results if the tests are run in a different order, which makes me a little skeptical of using that.

I wrote a very simple benchmark: https://gist.github.com/1518620

Just put the URL of the script file as the hash. For example:
metamorph-bench.html#https://raw.github.com/tomhuda/metamorph.js/master/lib/metamorph.js
metamorph-bench.html#https://raw.github.com/nateps/metamorph.js/master/lib/metamorph.js

Based on this benchmark, the comments implementation is now about the same speed. You can compare to the old implementation:
metamorph-bench.html#https://raw.github.com/nateps/metamorph.js/f3346a5957024bd0feca336bd226844fdd899f41/lib/metamorph.js

In FF9, old comment implementation took 4433ms, new comment version took 1756ms, and script version took 1780ms.

nateps commented:
Unfortunately, I ran into a hiccup with IE9 and `<select>` elements that I haven't figured out how to resolve. I hadn't noticed before, because IE9 was not actually using the range implementation, and it was falling back to `<script>` tags in my previous commits.

Based on reading the code, I had thought that IE9 was using the range code, since there was a shim for Range.prototype.createContextualFragment. However, that implementation didn't handle contents that need to be wrapped, and IE9 was using the non-range implementation. I fixed the Range.prototype.createContextualFragment shim and refactored the IE code to use the same fragment generator (which is more DRY and should be faster to boot).

The problem is that IE9 does not support the `<comment>` elements being used for IE6-8, and like IE6-8, IE9 removes comments that are children of `<select>` elements when it parses HTML. With the comment elements, I had figured out that IE would still insert the contents of `<comment>` elements as non-rendered text nodes. IE9 will also leave text nodes in place and doesn't render them within a select. However, the text node has to just be text, and it can't be wrapped in a comment. Thus, adding markers within selects would require different wrapper nodes.

For IE9, it would be possible to implement the use of comment markers in most places and text nodes markers in select elements, but it would require having the developer specifically note whether the morph is going to be inserted in a select or not. Certainly not ideal, but that's the best idea I have come up with so far that would use comments for all browsers and work with IE9.

Do people think it makes sense to add a "forSelect" boolean argument to morph.outerHTML that would return text node markers in IE9?

nateps committed:
Add "forSelect" argument to morph.outerHTML to address IE9 bug with comment markers in select elements

nateps committed:
New fix for IE9 markers in select that does not require API changes

In IE9, return script markers from morph.outerHTML, and then replace
them with comments as soon as they are created in the DOM by handling
the DOMNodeInserted event.

nateps commented:
Came up with a much cleaner solution for IE9. If the browser strips comments that are within selects, morph.outerHTML will return script tag markers. However, as soon as the script elements are created, they are replaced with comments when the DOMNodeInserted event is fired. No special context awareness needed.

All tests pass in all browsers now.

nateps commented:
Any further questions or concerns about this pull request? As far as I know, all outstanding issues with it have been fixed.

nateps commented:
Ping. Any more comments / concerns?

Thanks!
Nate

ebryn commented:
@nateps I think there are still performance concerns with the iterator approach, especially in documents with a large number of elements.

devinrhode2 commented:
I ran a bunch of tests in a variety of browsers, for the limited test that nate provided and also for a super massive page. I've come to the same conclusions as nate, except for IE9. IE9 had double the processing time than the tomhuda/master version. This makes sense because scripts are initially used and then the page is re-processed to replace those with comments.

@nateps I think your approach to have a forSelect option works best. That's a more isolated quirk than literring the dom with script tags. Plus I think you can just see if it is in a <select> with JSdom in derby.

tomdale commented:
@nateps Are you using this approach in production? What browsers are you supporting? We continue to have performance reservations about this approach, and would prefer to work with the browser vendors to implement this feature as a first-class citizen.

devinrhode2 commented:
Here's a good showcase of testing data:

<a href="http://www.browserscope.org/user/tests/table/agt1YS1wcm9maWxlcnINCxIEVGVzdBiO6pgSDA?v=3&layout=simple" target="_blank"><img src="http://cl.ly/image/1l3O0s0L0q0V/Screen%20Shot%202012-09-24%20at%202.31.35%20PM.png" /><br>Check live data</a>
Somehow my Firefox 15 was logged as "Firefox 3.6.12" - this is why the times are soo good. "Firefox 3.6.17" is actually firefox 3.6 via browserstack.

Here's the test page:
http://thescoutapp.com/test/metamorph-browserstack.php?username=nateps

It alternates from testing nate's version and tomhuda's version.

I have another test setup with a massive DOM and the test results vary widely, but the doubling for IE9 is clear:
http://www.browserscope.org/user/tests/table/agt1YS1wcm9maWxlcnINCxIEVGVzdBjlx5kSDA?v=3&layout=simple

The test page for this: http://thescoutapp.com/test/twitter.php?username=nateps

SImply load the test page and let it sit. It will wait for window.onload, then runs the test code, which is the same as nate's but the number of`while`iterations is shortened for the 2nd massive test, otherwise running the test in IE would break with an unresponsive script.

I made sure to remove all scripts from twitter, so nothing else is happening`onload`

As the test page reloads you can see data populate data in the browserscope page.

devinrhode2 commented:
@tomdale what if there were some Ember-jQuery.mod.js that hacks node traversal, ember script tags would be ignored? This would definitely help the situation (but could totally throw off developers...)

Since IE9 has bad performance, I definitely back Tom's decision to stick to script tags. However, if there was a way to fix this performance issue, perhaps with some server side interaction or even a build hook (ew) it would undeniably be better. The data shows that they are on par with performance. The browser scope test above was with an incredibly massive dom.
